src/hooks/transactions/useSignTransactions.tsx

```diff
@@ -1,189 +1,236 @@
+      onCancel(errorMessage, sessionId);
+    
+    };
+    
+  };
+
+  const signTransactions = async () => {
+
+    const hasSessionId = Boolean(transactionsToSign?.sessionId);
+    const { sessionId, transactions, callbackRoute } = transactionsToSign!;
+    const hasTransactions = Boolean(transactionsToSign?.transactions?.length);
+    const isSignEligible = transactionsToSign && hasSessionId && hasTransactions;
+    //move to errors constants
+    const missingProviderErrorMessage = 'You need a signer/valid signer to send a transaction,use either WalletProvider, LedgerProvider or WalletConnect';
+
+    if (!isSignEligible) {
+      return;
     }
-  }
 
-  useEffect(() => {
-    if (transactionsToSign?.sessionId) {
-      signTransactions();
+    if (!provider) {
+      console.error(missingProviderErrorMessage);
+      return;
+    }
+
+   
+    /*
+    * if the transaction is cancelled 
+    * the callback will go to undefined,
+    * we save the most recent one for a valid transaction
+    */
+    savedCallback.current = callbackRoute || window.location.pathname;
+
+    const signWithWallet = ( transactions: Array<Transaction> ) => {
+
+      const urlParams = { [walletSignSession]: sessionId };
+      const callbackUrl = `${window.location.origin}${callbackRoute}`;
+      const buildedCallbackUrl = buildReplyUrl({ callbackUrl, urlParams });
+
+      provider.signTransactions(transactions, {
+        callbackUrl: encodeURIComponent(buildedCallbackUrl)
+      });
+      
+    };
+
+    const mapTransactionsNonces = ( latestNonce: number, transactions: Array<Transaction> ): Array<Transaction> => {
+
+      return transactions.map((tx: Transaction, index: number) => {
+        
+        tx.setNonce(new Nonce(latestNonce + index));
+
+        return(tx);
+
+      });
+
+    };
+
+    try {
+      const proxyAccount = await proxy.getAccount(new Address(address));  
+      const isSigningWithWebWallet = providerType === LoginMethodsEnum.wallet;
+      const latestNonce = getLatestNonce(proxyAccount);
+      const mappedTransactions = mapTransactionsNonces(
+        latestNonce,
+        transactions
+      );
+
+      if(isSigningWithWebWallet) {
+        
+        signWithWallet(mappedTransactions);
+    
+      } else {
+      
+        signTransactionsWithProvider();
+      
+      }
+    
+    } catch (err) {  
+    
+      const customErrorMessage = 'error when signing';
+      const defaultErrorMessage = (error as unknown as Error)?.message;
+      const errorMessage = customErrorMessage || defaultErrorMessage;
+      onCancel(errorMessage, sessionId);
+
+      dispatch(
+        moveTransactionsToSignedState({
+          sessionId,
+          status: TransactionBatchStatusesEnum.cancelled
+        })
+      );
+
+      console.error(errorMessage, err);
+
     }
-  }, [transactionsToSign?.sessionId]);
-  const hasTransactions = transactionsToSign?.transactions;
+  };
+
+  useEffect(() => {
+
+    signTransactions();
```

Comment:
what if transactionsToSign is null? This will go into signTransactions on crash on destructuring